Sporadic failure in test SecureBookKeeperLogTests.testAppendTransientBookieFailure

**Problem description**
Found the test `io.pravega.segmentstore.storage.impl.bookkeeper.BookKeeperLogTests$SecureBookKeeperLogTests.testAppendTransientBookieFailure` failing due to:
```
java.util.concurrent.CompletionException: io.pravega.common.util.RetriesExhaustedException: io.pravega.segmentstore.storage.WriteFailureException: Active Ledger is closed.
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:331)
	at java.base/java.util.concurrent.CompletableFuture.andTree(CompletableFuture.java:1452)
	at java.base/java.util.concurrent.CompletableFuture.andTree(CompletableFuture.java:1442)
	at java.base/java.util.concurrent.CompletableFuture.andTree(CompletableFuture.java:1442)
	at java.base/java.util.concurrent.CompletableFuture.andTree(CompletableFuture.java:1442)
	at java.base/java.util.concurrent.CompletableFuture.andTree(CompletableFuture.java:1442)
	at java.base/java.util.concurrent.CompletableFuture.andTree(CompletableFuture.java:1442)
	at java.base/java.util.concurrent.CompletableFuture.andTree(CompletableFuture.java:1442)
	at java.base/java.util.concurrent.CompletableFuture.andTree(CompletableFuture.java:1442)
	at java.base/java.util.concurrent.CompletableFuture.andTree(CompletableFuture.java:1442)
	at java.base/java.util.concurrent.CompletableFuture.allOf(CompletableFuture.java:2337)
	at io.pravega.common.concurrent.Futures.allOfWithResults(Futures.java:437)
	at io.pravega.segmentstore.storage.impl.bookkeeper.BookKeeperLogTests.testAppendTransientBookieFailure(BookKeeperLogTests.java:274)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: io.pravega.common.util.RetriesExhaustedException: io.pravega.segmentstore.storage.WriteFailureException: Active Ledger is closed.
	at io.pravega.segmentstore.storage.impl.bookkeeper.BookKeeperLog.executeWrites(BookKeeperLog.java:481)
	at io.pravega.segmentstore.storage.impl.bookkeeper.BookKeeperLog.processPendingWrites(BookKeeperLog.java:430)
	at io.pravega.segmentstore.storage.impl.bookkeeper.BookKeeperLog.processWritesSync(BookKeeperLog.java:390)
	at java.base/java.util.concurrent.CompletableFuture$UniRun.tryFire(CompletableFuture.java:783)
	at java.base/java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:478)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	... 1 more
Caused by: io.pravega.segmentstore.storage.WriteFailureException: Active Ledger is closed.
	at io.pravega.segmentstore.storage.impl.bookkeeper.BookKeeperLog.handleWriteException(BookKeeperLog.java:668)
	at io.pravega.segmentstore.storage.impl.bookkeeper.BookKeeperLog.addCallback(BookKeeperLog.java:602)
	at io.pravega.segmentstore.storage.impl.bookkeeper.BookKeeperLog.lambda$executeWrites$5(BookKeeperLog.java:488)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2088)
	at org.apache.bookkeeper.client.SyncCallbackUtils.finish(SyncCallbackUtils.java:83)
	at org.apache.bookkeeper.client.SyncCallbackUtils$SyncAddCallback.addComplete(SyncCallbackUtils.java:251)
	at org.apache.bookkeeper.client.AsyncCallback$AddCallback.addCompleteWithLatency(AsyncCallback.java:92)
	at org.apache.bookkeeper.client.LedgerHandle$9.safeRun(LedgerHandle.java:1324)
	at org.apache.bookkeeper.common.util.SafeRunnable.run(SafeRunnable.java:36)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	... 1 more
Caused by: org.apache.bookkeeper.client.BKException$BKLedgerClosedException: Attempt to write to a closed ledger
	... 9 more
	```

This test has been fixed in the past, but it seems that it is still flaky.

**Problem location**
`io.pravega.segmentstore.storage.impl.bookkeeper.BookKeeperLogTests$SecureBookKeeperLogTests.testAppendTransientBookieFailure`

**Suggestions for an improvement**
Make the test to pass reliably.

--- a/segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperServiceRunner.java
+++ b/segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperServiceRunner.java
@@ -115,6 +115,34 @@ public class BookKeeperServiceRunner implements AutoCloseable {
     }
 
     /**
+     * Suspends processing for the BookieService with the given index.
+     *
+     * @param bookieIndex The index of the bookie to stop.
+     */
+    public void suspendBookie(int bookieIndex) {
+        Preconditions.checkState(this.servers.size() > 0, "No Bookies initialized. Call startAll().");
+        Preconditions.checkState(this.servers.get(0) != null, "Bookie does not exists.");
+        val bk = this.servers.get(bookieIndex);
+        log.info("Bookie {} is suspending processing.", bookieIndex);
+        bk.suspendProcessing();
+        log.info("Bookie {} suspended processing.", bookieIndex);
+    }
+
+    /**
+     * Resumes processing for the BookieService with the given index.
+     *
+     * @param bookieIndex The index of the bookie to stop.
+     */
+    public void resumeBookie(int bookieIndex) {
+        Preconditions.checkState(this.servers.size() > 0, "No Bookies initialized. Call startAll().");
+        Preconditions.checkState(this.servers.get(0) != null, "Bookie does not exists.");
+        val bk = this.servers.get(bookieIndex);
+        log.info("Bookie {} is resuming processing.", bookieIndex);
+        bk.resumeProcessing();
+        log.info("Bookie {} resumed processing.", bookieIndex);
+    }
+
+    /**
      * Restarts the BookieService with the given index.
      *
      * @param bookieIndex The index of the bookie to restart.

--- a/segmentstore/storage/impl/src/test/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperLogTests.java
+++ b/segmentstore/storage/impl/src/test/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperLogTests.java
@@ -268,6 +268,8 @@ public abstract class BookKeeperLogTests extends DurableDataLogTestBase {
             } finally {
                 // Resume the bookie with the appends still in flight.
                 resumeFirstBookie();
+                AssertExtensions.assertThrows("Bookies should be running, but they aren't",
+                        BookKeeperLogTests::restartFirstBookie, ex -> ex instanceof IllegalStateException);
             }
 
             // Wait for all writes to complete, then reassemble the data in the order set by LogAddress.
@@ -319,6 +321,10 @@ public abstract class BookKeeperLogTests extends DurableDataLogTestBase {
                                 }
                             });
                 }
+                AssertExtensions.assertThrows("Bookies shouldn't be running, but they are",
+                        BookKeeperLogTests::suspendFirstBookie, ex -> ex instanceof IllegalStateException);
+                AssertExtensions.assertThrows("Bookies shouldn't be running, but they are",
+                        BookKeeperLogTests::resumeFirstBookie, ex -> ex instanceof IllegalStateException);
             } finally {
                 // Don't forget to resume the bookie, but only AFTER we are done testing.
                 restartFirstBookie();

--- a/segmentstore/storage/impl/src/test/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperLogTests.java
+++ b/segmentstore/storage/impl/src/test/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperLogTests.java
@@ -256,7 +256,7 @@ public abstract class BookKeeperLogTests extends DurableDataLogTestBase {
 
             try {
                 // Suspend a bookie (this will trigger write errors).
-                stopFirstBookie();
+                suspendFirstBookie();
 
                 // Issue appends in parallel, without waiting for them.
                 int writeCount = getWriteCount();
@@ -267,7 +267,7 @@ public abstract class BookKeeperLogTests extends DurableDataLogTestBase {
                 }
             } finally {
                 // Resume the bookie with the appends still in flight.
-                restartFirstBookie();
+                resumeFirstBookie();
             }
 
             // Wait for all writes to complete, then reassemble the data in the order set by LogAddress.
@@ -768,6 +768,14 @@ public abstract class BookKeeperLogTests extends DurableDataLogTestBase {
 
     private static void stopFirstBookie() {
         BK_SERVICE.get().stopBookie(0);
+    }
+
+    private static void suspendFirstBookie() {
+        BK_SERVICE.get().suspendBookie(0);
+    }
+
+    private static void resumeFirstBookie() {
+        BK_SERVICE.get().resumeBookie(0);
     }
 
     @SneakyThrows

